### ISSUE TITLE
Stryker 4.4.0 no longer runs in the Github Build Pipeline

### ISSUE THREAD

**Issue by vbreuss** [stryker-mutator/stryker-net#3103]
**Describe the bug**
I include Stryker in my build pipeline and it is suddenly failing, after the update to 4.4.0:
https://github.com/Testably/Testably.Expectations/actions/runs/11749239517/job/32735250575

**Logs**
The "normal" logs are:
```
_____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.4.0

[20:46:36 INF] Analysis starting.
[20:46:36 INF] Analyzing 7 test project(s).
[20:46:57 INF] Found project /home/runner/work/Testably.Expectations/Testably.Expectations/Source/Testably.Expectations/Testably.Expectations.csproj to mutate.
[20:46:58 INF] Analysis complete.
[20:46:58 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.XUnit2.Tests/TestFramework.XUnit2.Tests.csproj (1/7)
[20:46:58 INF] Building project TestFramework.XUnit2.Tests.csproj using dotnet build TestFramework.XUnit2.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.XUnit2.Tests.)
[20:47:21 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.NUnit4.Tests/TestFramework.NUnit4.Tests.csproj (2/7)
[20:47:21 INF] Building project TestFramework.NUnit4.Tests.csproj using dotnet build TestFramework.NUnit4.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.NUnit4.Tests.)
[20:47:25 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.FallbackTest/TestFramework.FallbackTest.csproj (3/7)
[20:47:25 INF] Building project TestFramework.FallbackTest.csproj using dotnet build TestFramework.FallbackTest.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.FallbackTest.)
[20:47:28 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/Testably.Expectations.Internal.Tests/Testably.Expectations.Internal.Tests.csproj (4/7)
[20:47:28 INF] Building project Testably.Expectations.Internal.Tests.csproj using dotnet build Testably.Expectations.Internal.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/Testably.Expectations.Internal.Tests.)
[20:47:32 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.TUnit.Tests/TestFramework.TUnit.Tests.csproj (5/7)
[20:47:32 INF] Building project TestFramework.TUnit.Tests.csproj using dotnet build TestFramework.TUnit.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.TUnit.Tests.)
[20:47:35 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.NUnit3.Tests/TestFramework.NUnit3.Tests.csproj (6/7)
[20:47:35 INF] Building project TestFramework.NUnit3.Tests.csproj using dotnet build TestFramework.NUnit3.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/TestFrameworks/TestFramework.NUnit3.Tests.)
[20:47:38 INF] Building test project /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/Testably.Expectations.Tests/Testably.Expectations.Tests.csproj (7/7)
[20:47:38 INF] Building project Testably.Expectations.Tests.csproj using dotnet build Testably.Expectations.Tests.csproj (directory /home/runner/work/Testably.Expectations/Testably.Expectations/Tests/Testably.Expectations.Tests.)
[20:47:49 ERR] An error occurred during the mutation test run 
System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at Stryker.Core.Helpers.VsTestHelper.DeployEmbeddedVsTestBinaries()
   at Stryker.Core.Helpers.VsTestHelper.GetVsTestToolPaths()
   at Stryker.Core.Helpers.VsTestHelper.GetCurrentPlatformVsTestToolPath()
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.BuildActualVsTestWrapper(ConsoleParameters parameters)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.BuildVsTestWrapper(String runnerId)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.DiscoverTestsInSources(String newSource, String frameworkVersion, String platform)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.AddTestSource(String source, String frameworkVersion, String platform)
   at Stryker.Core.TestRunners.VsTest.VsTestRunnerPool.DiscoverTests(String assembly)
   at Stryker.Core.Initialisation.InitialisationProcess.DiscoverTests(SourceProjectInfo projectInfo, ITestRunner testRunner)
   at Stryker.Core.Initialisation.InitialisationProcess.InitialTest(IStrykerOptions options, SourceProjectInfo projectInfo, ITestRunner testRunner, Boolean throwIfFails)
   at Stryker.Core.Initialisation.InitialisationProcess.GetMutationTestInputs(IStrykerOptions options, IReadOnlyCollection`1 projects, ITestRunner runner)
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(IStrykerOptions options, IReporter reporters, ITestRunner runner)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator)
[20:47:49 INF] Time Elapsed 00:01:14.1965592
Unhandled exception. System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at Stryker.Core.Helpers.VsTestHelper.DeployEmbeddedVsTestBinaries()
   at Stryker.Core.Helpers.VsTestHelper.GetVsTestToolPaths()
   at Stryker.Core.Helpers.VsTestHelper.GetCurrentPlatformVsTestToolPath()
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.BuildActualVsTestWrapper(ConsoleParameters parameters)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.BuildVsTestWrapper(String runnerId)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.DiscoverTestsInSources(String newSource, String frameworkVersion, String platform)
   at Stryker.Core.TestRunners.VsTest.VsTestContextInformation.AddTestSource(String source, String frameworkVersion, String platform)
   at Stryker.Core.TestRunners.VsTest.VsTestRunnerPool.DiscoverTests(String assembly)
   at Stryker.Core.Initialisation.InitialisationProcess.DiscoverTests(SourceProjectInfo projectInfo, ITestRunner testRunner)
   at Stryker.Core.Initialisation.InitialisationProcess.InitialTest(IStrykerOptions options, SourceProjectInfo projectInfo, ITestRunner testRunner, Boolean throwIfFails)
   at Stryker.Core.Initialisation.InitialisationProcess.GetMutationTestInputs(IStrykerOptions options, IReadOnlyCollection`1 projects, ITestRunner runner)
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(IStrykerOptions options, IReporter reporters, ITestRunner runner)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator)
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 152
   at Stryker.CLI.StrykerCli.<>c__DisplayClass[11](https://github.com/Testably/Testably.Expectations/actions/runs/11749239517/job/32735250575#step:6:12)_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line [12](https://github.com/Testably/Testably.Expectations/actions/runs/11749239517/job/32735250575#step:6:13)4
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass143_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line [13](https://github.com/Testably/Testably.Expectations/actions/runs/11749239517/job/32735250575#step:6:14)0
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line [14](https://github.com/Testably/Testably.Expectations/actions/runs/11749239517/job/32735250575#step:6:15)
/home/runner/work/_temp/7f4cee0e-db3d-4303-a33f-5a339896d3ba.sh: line 2:  2060 Aborted                 (core dumped) ../../tools/dotnet-stryker -f ../.github/stryker/Stryker.CI.Config.json -v "${GITHUB_HEAD_REF}" -r "Dashboard" -r "html" -r "cleartext" --since:main
Error: Process completed with exit code 134.
```

As it is part of the build pipeline, I am not sure how to easily create more detailed logs.

**Expected behavior**
The pipeline should succeed

**Desktop (please complete the following information):**
 - OS: Ubuntu (Github Runner)
 - Type of project [core]
 - Framework Version [net 6 and net 8]
 - Stryker Version [4.4.0]

**Additional context**
The relevant part of the yml pipeline:
```yml
    stryker:
        name: Stryker mutation testing
        runs-on: ubuntu-latest
        timeout-minutes: 300
        steps:
            - name: Checkout sources
              uses: actions/checkout@v4
              with:
                  fetch-depth: 0
            - name: Setup .NET
              uses: actions/setup-dotnet@v4
              with:
                  dotnet-version: |
                      6.0.x
                      8.0.x
            - name: Install .NET Stryker
              shell: bash
              run: |
                  dotnet tool install dotnet-stryker --tool-path ../tools
            - name: Prepare Reports directory
              shell: bash
              run: |
                  mkdir Tests/StrykerOutput/Reports -p
            - name: Analyze Testably.Expectations
              env:
                  STRYKER_DASHBOARD_API_KEY: ${{ secrets.STRYKER_DASHBOARD_API_KEY }}
              shell: bash
              run: |
                  cd Tests
                  ../../tools/dotnet-stryker -f ../.github/stryker/Stryker.CI.Config.json -v "${GITHUB_HEAD_REF}" -r "Dashboard" -r "html" -r "cleartext" --since:main
                  mv ./StrykerOutput/**/reports/*.html ./StrykerOutput/Reports/Testably.Expectations-report.html
```

**Comment by vbreuss** [comment]
In my other Project ([Testably.Abstractions](https://github.com/Testably/Testably.Abstractions/actions/runs/11749265949/job/32735125757?pr=664)) I run Stryker both on Ubuntu and Windows and the Windows-Runner seems to work, so probably the issue is related to Ubuntu?!?

**Comment by rouke-broersma** [comment]
We worked around some concurrency issues in the latest release, which seems to have uncovered some other concurrency issues.. Thanks for reporting!

**Comment by 0xced** [comment]
I've been hit by this too and did some investigation. It turns out you can easily reproduce this issue by deleting the `~/.nuget/packages/microsoft.testplatform.portable` directory before running `dotnet stryker`.

When `vstest.console.dll` or `vstest.console.exe` is not found in the `~/.nuget/packages/microsoft.testplatform.portable` directory, Stryker tries to deploy its embedded version. But it has none!

The exception comes from here: https://github.com/stryker-mutator/stryker-net/blob/957cf460b1a30509fe394dd780f174bceac9f145/src/Stryker.Core/Stryker.Core/Helpers/VsTestHelper.cs#L189-L192

When that feature was introduced, in _feat(Dotnet Framework support): Add initial support for dotnet full framework (#371)_, the `microsoft.testplatform.portable` nupkg file was embedded into the `Stryker.Core` package: https://github.com/stryker-mutator/stryker-net/blob/93ca8a5f37e76885ded1c2f4cb7c12123081cc9d/src/Stryker.Core/Stryker.Core/Stryker.Core.csproj#L28

Note that downloading of this package happens at build time: https://github.com/stryker-mutator/stryker-net/blob/957cf460b1a30509fe394dd780f174bceac9f145/azure-pipelines.yml#L137-L142

But this line responsible for embedding the nupkg was deleted in 0aeb7bc6fc9f9bc5366835c0b2ddc3ab46f1f52f, which looks like the culprit!

**Comment by rouke-broersma** [comment]
Oh boy that's not good.. Thanks for investigating!

**Comment by 0xced** [comment]
I'm attempting a fix with a simplified workflow to embed the nupkg. Will submit a pull request soon.

**Comment by kevinpearce** [comment]
The same seen on Azure DevOps agents 
- 4.3.0 == OK, 
- 4.4.0 == As above, "System.InvalidOperationException: Sequence contains no matching element"

**Comment by vbreuss** [comment]
Unfortunately I still got an error with 4.4.1:
```
Run cd Tests
  cd Tests
  ../../tools/dotnet-stryker -f ../.github/stryker/Stryker.Config.Testing.json -v "${GITHUB_HEAD_REF}" -r "Dashboard" -r "html" -r "cleartext" --since:main
  mv ./StrykerOutput/**/reports/*.html ./StrykerOutput/Reports/Testably.Abstractions.Testing-report.html
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    DOTNET_ROOT: /usr/share/dotnet
    STRYKER_DASHBOARD_API_KEY: ***

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.4.1

[19:31:56 INF] Analysis starting.
[19:31:56 INF] Analyzing 2 test project(s).
[19:32:04 INF] Analysis complete.
[19:32:04 ERR] An error occurred during the mutation test run 
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at Stryker.Core.Initialisation.InputFileResolver.AnalyzeAllNeededProjects(List`1 projectList, IStrykerOptions options, IAnalyzerManager manager, ScanMode mode)
   at Stryker.Core.Initialisation.InputFileResolver.AnalyzeAndIdentifyProjects(List`1 projectList, IStrykerOptions options, IAnalyzerManager manager, ScanMode mode)
   at Stryker.Core.Initialisation.InputFileResolver.ResolveSourceProjectInfos(IStrykerOptions options)
   at Stryker.Core.Initialisation.InitialisationProcess.GetMutableProjectsInfo(IStrykerOptions options)
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(IStrykerOptions options, IReporter reporters, ITestRunner runner)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator)
[19:32:04 INF] Time Elapsed 00:00:09.0750260
Unhandled exception. System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at Stryker.Core.Initialisation.InputFileResolver.AnalyzeAllNeededProjects(List`1 projectList, IStrykerOptions options, IAnalyzerManager manager, ScanMode mode)
   at Stryker.Core.Initialisation.InputFileResolver.AnalyzeAndIdentifyProjects(List`1 projectList, IStrykerOptions options, IAnalyzerManager manager, ScanMode mode)
   at Stryker.Core.Initialisation.InputFileResolver.ResolveSourceProjectInfos(IStrykerOptions options)
   at Stryker.Core.Initialisation.InitialisationProcess.GetMutableProjectsInfo(IStrykerOptions options)
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(IStrykerOptions options, IReporter reporters, ITestRunner runner)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator)
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 152
   at Stryker.CLI.StrykerCli.<>c__DisplayClass11_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 124
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass143_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 130
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line 14
/home/runner/work/_temp/5b14c011-54cb-4878-9862-bf600e6ec71f.sh: line 2:  2200 Aborted                 (core dumped) ../../tools/dotnet-stryker -f ../.github/stryker/Stryker.Config.Testing.json -v "${GITHUB_HEAD_REF}" -r "Dashboard" -r "html" -r "cleartext" --since:main
Error: Process completed with exit code 134.
```

**Comment by dupdob** [comment]
I suppose you can use `--dev-mode` as a work around: it will disable concurrent project analysis (which looks like the cause of problem here).
But there is nothing new there. The issues you faced so far are not new, so it is unclear why they appear with V4.4

**Comment by 0xced** [comment]
I'd suggest opening a new issue about your problem because it's definitely different from your initial post. The `System.InvalidOperationException: Sequence contains no matching element` error that you were experiencing has been fixed in version 4.4.1.

**Comment by dupdob** [comment]
For info: I opened a new issue with @vbreuss message, identified and fixed the issue. See PR #3109
